ord_schema/process_dataset.py

```diff
@@ -0,0 +1,143 @@
+"""Dataset processing script for database submissions.
+
+This script is meant to be a one-stop shop for preparing submissions to the
+Open Reaction Database.
+
+By default, the script only validates the input Dataset messages. Validation
+may introduce changes to the Reaction messages, such as the addition of SMILES
+for compounds identified only by NAME. Users should frequently run these checks
+as they are preparing a dataset for submission.
+
+With the optional --update flag, the script also performs database-specific
+updates (such as adding record IDs). These operations are meant to be run as
+part of the submission process and not as part of the pre-submission validation
+cycle.
+
+Example usage:
+* For normal validation-only operation:
+  $ python process_dataset.py --input_pattern=my_dataset.pb
+* To write the validated protos to disk:
+  $ python process_dataset.py --input_pattern=my_dataset.pb --output=out.pb
+* To write errors to disk:
+  $ python process_dataset.py --input_pattern=my_dataset.pb --write_errors
+* To process multiple Dataset protos:
+  $ python process_dataset.py --input_pattern="my_dataset-*.pb"
+"""
+
+import glob
+import os
+import uuid
+
+from absl import app
+from absl import flags
+from absl import logging
+
+from ord_schema import message_helpers
+from ord_schema import validations
+from ord_schema.proto import dataset_pb2
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string('input_pattern', None,
+                    'Pattern (glob) matching input Dataset protos.')
+flags.DEFINE_enum('input_format', 'binary',
+                  [f.value for f in message_helpers.MessageFormats],
+                  'Input message format.')
+flags.DEFINE_boolean('write_errors', False,
+                     'If True, errors will be written to <filename>.error.')
+flags.DEFINE_string('output', None, 'Filename for output Dataset.')
+flags.DEFINE_boolean('validate', True, 'If True, validate Reaction protos.')
+flags.DEFINE_boolean('update', False, 'If True, update Reaction protos.')
+
+
+def validate(datasets):
+    """Runs validation for a set of datasets.
+
+    Args:
+        datasets: Dict mapping filenames to Dataset protos.
+    """
+    validation_errors = False
+    for filename, dataset in datasets.items():
+        errors = _validate_dataset(filename, dataset)
+        if errors:
+            validation_errors = True
+            if FLAGS.write_errors:
+                with open(f'{filename}.error', 'w') as f:
+                    for error in errors:
+                        f.write(f'{error}\n')
+    # NOTE(kearnes): Run validation for all datasets before
+    # exiting if there are errors.
+    if validation_errors:
+        raise ValueError('validation encountered errors')
```

Comment:
Is it worth expanding this to something like "validation encountered errors; re-run with '--write_errors true' to save errors to file"? While this script should only be seen by superusers/developers, might make it easier to remember this option.

Reply:
By default all the errors will be written to the console, so they're not hidden. Do you still want that suggestion in that case?

Reply:
In that case, no. What you have makes sense.

Reply:
Added.

Reply:
One interesting note about boolean flags: if a flag defaults to False, you only need to pass the flag to turn it on: `--write_errors`. If a flag defaults to True, the flag module creates a special flag to turn it off: `--nowrite_errors` (just the flag name prefixed with `no`).